segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
                             ChunkedSegmentStorageConfig config,
                             ScheduledExecutorService storageExecutor,
                             Supplier<Long> currentTimeSupplier,
-                            Supplier<CompletableFuture<Void>> delaySupplier) {
-        super(String.format("GarbageCollector[%d]", containerId), ExecutorServiceHelpers.newScheduledThreadPool(1, "storage-gc"));
-        try {
-            this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
-            this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
-            this.config = Preconditions.checkNotNull(config, "config");
-            this.currentTimeSupplier = Preconditions.checkNotNull(currentTimeSupplier, "currentTimeSupplier");
-            this.delaySupplier = Preconditions.checkNotNull(delaySupplier, "delaySupplier");
-            this.storageExecutor = Preconditions.checkNotNull(storageExecutor, "storageExecutor");
-        } catch (Exception ex) {
-            this.executor.shutdownNow();
-            throw ex;
-        }
+                            Function<Duration, CompletableFuture<Void>> delaySupplier) {
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.currentTimeSupplier = Preconditions.checkNotNull(currentTimeSupplier, "currentTimeSupplier");
+        this.delaySupplier = Preconditions.checkNotNull(delaySupplier, "delaySupplier");
+        this.storageExecutor = Preconditions.checkNotNull(storageExecutor, "storageExecutor");
+        this.traceObjectId = String.format("GarbageCollector[%d]", containerId);
+        this.taskQueueName = String.format("GC.queue.%d", containerId);
+        this.failedQueueName = String.format("GC.failed.queue.%d", containerId);
+        this.taskSchedular = new MultiKeySequentialProcessor<>(storageExecutor);
     }
 
     /**
      * Initializes this instance.
+     * @param taskQueue Task queue to use.
      */
-    public void initialize() {
-        Services.startAsync(this, this.executor);
+    public CompletableFuture<Void> initialize(AbstractTaskQueue<TaskInfo> taskQueue) {
+        // Temp change to be removed after next PR (part 2 of 3)
```

Comment:
TODO with Github Issue

Please write TODOs with Github issue(s) below as well - I see you used similar patterns to disable features.

Reply:
I'll be combining PR part 1 and 2 , now that Raul's changes are in.

Reply:
merged 2 PRs